uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -8,68 +8,26 @@
 //!
 //! # The big picture
 //!
-//! What happens when you call an async function exported from the Rust API?
+//! We implement async foreign functions using a simplified version of the Future API:
 //!
-//! 1. You make a call to a generated async function in the foreign bindings.
-//! 2. That function suspends itself, then makes a scaffolding call.  In addition to the normal
-//!    arguments, it passes a `ForeignExecutor` and callback.
-//! 3. Rust uses the `ForeignExecutor` to schedules polls of the Future until it's ready.  Then
-//!    invokes the callback.
-//! 4. The callback resumes the suspended async function from (2), closing the loop.
+//! 1. Call the scaffolding function to get a [RustFutureHandle]
+//! 2a. In a loop:
+//!   - Call [rust_future_poll]
+//!   - Suspend the function until the [rust_future_poll] continuation function is called
+//!   - If the continuation was function was called with [RustFuturePoll::Ready], then break
+//!     otherwise continue.
+//! 2b. If the async function is cancelled, then call [rust_future_cancel].  This causes the
+//!     continuation function to be called with [RustFuturePoll::Ready] and the [RustFuture] to
+//!     enter a cancelled state.
+//! 3. Call [rust_future_complete] to get the result of the future.
+//! 4. Call [rust_future_free] to free the future, ideally in a finally block.  This:
+//!    - Releases any resources held by the future
+//!    - Calls any continuation callbacks that have not been called yet
```

Comment:
This PR could be called @Hywan was right.  This new API is basically the one he initially proposed with just some tweaks to simplify the FFI and bindings code.

Reply:
Thanks for recognizing it 👼.